Signal/src/Models/TSMessageAdapaters/TSGenericAttachmentAdapter.m

```diff
@@ -131,35 +133,66 @@ - (UIView *)mediaView
             = round(iconFrame.origin.x + (iconFrame.size.width - fileTypeLabelFrame.size.width) * 0.5f);
         fileTypeLabelFrame.origin.y
-            = round(iconFrame.origin.y + (iconFrame.size.height - fileTypeLabelFrame.size.height) * 0.5f + 5);
+            = round(iconFrame.origin.y + (iconFrame.size.height - fileTypeLabelFrame.size.height) * 0.5f);
         fileTypeLabel.frame = fileTypeLabelFrame;
         [_cachedMediaView addSubview:fileTypeLabel];
 
-        UILabel *attachmentLabel = [UILabel new];
-        attachmentLabel.text = NSLocalizedString(@"GENERIC_ATTACHMENT_LABEL", @"A label for generic attachments.");
-        attachmentLabel.textColor = [textColor colorWithAlphaComponent:0.85f];
-        attachmentLabel.lineBreakMode = NSLineBreakByTruncatingTail;
-        attachmentLabel.font = [self attachmentLabelFont];
-        [attachmentLabel sizeToFit];
-        CGRect attachmentLabelFrame = CGRectZero;
-        attachmentLabelFrame.size = attachmentLabel.bounds.size;
-        attachmentLabelFrame.origin.x
-            = round(contentFrame.origin.x + (contentFrame.size.width - attachmentLabelFrame.size.width) * 0.5f);
-        attachmentLabelFrame.origin.y
-            = round(contentFrame.origin.y + contentFrame.size.height - attachmentLabelFrame.size.height);
-        attachmentLabel.frame = attachmentLabelFrame;
-        [_cachedMediaView addSubview:attachmentLabel];
+        const CGFloat kLabelHSpacing = 3;
+        const CGFloat kLabelVSpacing = 2;
+        NSString *topText =
+            [self.attachment.filename stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
+        if (topText.length < 1) {
+            topText = [MIMETypeUtil fileExtensionForMIMEType:self.attachment.contentType].uppercaseString;
+        }
+        if (topText.length < 1) {
+            topText = NSLocalizedString(@"GENERIC_ATTACHMENT_LABEL", @"A label for generic attachments.");
+        }
+        UILabel *topLabel = [UILabel new];
+        topLabel.text = topText;
+        topLabel.textColor = textColor;
+        topLabel.lineBreakMode = NSLineBreakByTruncatingMiddle;
+        topLabel.font = [UIFont ows_mediumFontWithSize:15.f];
+        [topLabel sizeToFit];
+        [_cachedMediaView addSubview:topLabel];
+
+        NSError *error;
+        unsigned long long fileSize =
+            [[NSFileManager defaultManager] attributesOfItemAtPath:self.attachment.filePath error:&error].fileSize;
+        OWSAssert(!error);
+        NSString *bottomText = [ViewControllerUtils formatFileSize:fileSize];
+        UILabel *bottomLabel = [UILabel new];
+        bottomLabel.text = bottomText;
+        bottomLabel.textColor = [textColor colorWithAlphaComponent:0.85f];
+        bottomLabel.lineBreakMode = NSLineBreakByTruncatingMiddle;
+        bottomLabel.font = [UIFont ows_regularFontWithSize:13.f];
+        [bottomLabel sizeToFit];
+        [_cachedMediaView addSubview:bottomLabel];
+
+        CGRect topLabelFrame = CGRectZero;
```

Comment:
This layout code is ugly, but I've made it more parameterized for easier tweaking.